// node-jq/lib/options.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var optionDefaults = exports.optionDefaults = {
  input: 'file',
  output: 'pretty',
  slurp: false,
  sort: false,
  raw: false
};

var optionMap = {
  output: {
    buildParams: function buildParams(params, value) {
      if (value === 'string' || value === 'compact') {
        params.unshift('--compact-output');
      }
    }
  },
  slurp: {
    buildParams: function buildParams(params, value) {
      if (value === true) {
        params.unshift('--slurp');
      }
    }
  },
  sort: {
    buildParams: function buildParams(params, value) {
      if (value === true) {
        params.unshift('--sort-keys');
      }
    }
  },
  color: {
    buildParams: function buildParams(params, value) {
      if (value === true) {
        params.unshift('--color-output');
      }
    }
  },
  raw: {
    buildParams: function buildParams(params, value) {
      if (value === true) {
        params.unshift('-r');
      }
    }
  }
};

var parseOptions = exports.parseOptions = function parseOptions() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  return Object.keys(options).reduce(function (params, key, index) {
    if (optionMap[key] !== undefined) {
      optionMap[key].buildParams(params, options[key]);
    }
    return params;
  }, []);
};
